### PR TITLE
Add OpenMP support to DenseBins::build

### PR DESCRIPTION
**Pull request by atmyers**
The proposed changes:
- [ ] fix a bug or incorrect behavior in AMReX
- [x] add new capabilities to AMReX
- [ ] changes answers in the test suite to more than roundoff level
- [ ] are likely to significantly affect the results of downstream AMReX users
- [ ] are described in the proposed changes to the AMReX documentation, if appropriate

### CODE REVIEW

**Comment by atmyers** [comment]
@MaxThevenet reports that this gives the correct results with Hipace++, and does accelerate building the bins for OpenMP runs:

```
output_newE_02.txt:DenseBins<T>::buildOpenMP                              772      3.558      3.558      3.558   2.02%
output_newE_04.txt:DenseBins<T>::buildOpenMP                              772      1.862      1.862      1.862   2.02%
output_newE_08.txt:DenseBins<T>::buildOpenMP                              772      1.086      1.086      1.086   2.16%
output_newE_16.txt:DenseBins<T>::buildOpenMP                              772       0.86       0.86       0.86   2.05%
output_oldE_02.txt:DenseBins<T>::build                                    772      6.816      6.816      6.816   3.79%
output_oldE_04.txt:DenseBins<T>::build                                    772      7.122      7.122      7.122   7.33%
output_oldE_08.txt:DenseBins<T>::build                                    772      7.065      7.065      7.065  12.39%
output_oldE_16.txt:DenseBins<T>::build                                    772      6.393      6.393      6.393  13.79%
```

File: Src/Particle/AMReX_DenseBins.H [hunk 1]
```diff
@@ -146,13 +223,250 @@ public:
...
+     * \param bx the Box that defines the space over which the bins will be defined
+     * \param f a function object that maps items to bins
+     */
+    template <typename N, typename F>
+    void build (BinPolicy::SerialBinPolicy, N nitems, T const* v, const Box& bx, F&& f)
+    {
+        const auto lo = lbound(bx);
+        const auto hi = ubound(bx);
+        build(BinPolicy::Serial, nitems, v, bx.numPts(),
+              [=] AMREX_GPU_DEVICE (const T& t) noexcept
```
Comment (WeiqunZhang) [inline Src/Particle/AMReX_DenseBins.H]: Why is it marked with `AMREX_GPU_DEVICE` for the serial case?

File: Src/Particle/AMReX_DenseBins.H [hunk 2]
```diff
@@ -146,13 +223,250 @@ public:
...
+     * \param bx the Box that defines the space over which the bins will be defined
+     * \param f a function object that maps items to bins
+     */
+    template <typename N, typename F>
+    void build (BinPolicy::OpenMPBinPolicy, N nitems, T const* v, const Box& bx, F&& f)
+    {
+        const auto lo = lbound(bx);
+        const auto hi = ubound(bx);
+        build(BinPolicy::OpenMP, nitems, v, bx.numPts(),
+              [=] AMREX_GPU_DEVICE (const T& t) noexcept
```
Comment (WeiqunZhang) [inline Src/Particle/AMReX_DenseBins.H]: Why is it marked `AMREX_GPU_DEVICE` for OpenMP?

File: Src/Particle/AMReX_DenseBins.H (hunk 1 above)
Comment (atmyers) [reply Src/Particle/AMReX_DenseBins.H]: Just come copy / paste gone wrong. Removed.

**Review by WeiqunZhang — APPROVED** [review]

**Review by jmsexton03 — APPROVED** [review]

**Pull request merged**